tests/bsim/bluetooth/host/misc/hfc/src/main.c

```diff
@@ -0,0 +1,444 @@
+{
+	int err;
+	static struct bt_gatt_subscribe_params params = {};
+
+	params.notify = cb;
+	params.subscribe = subscribed;
+	params.value = BT_GATT_CCC_NOTIFY;
+	params.value_handle = handle;
+	params.ccc_handle = handle + 1;
+
+	err = bt_gatt_subscribe(conn, &params);
+	ASSERT(!err, "Subscribe failed (err %d)\n", err);
+
+	WAIT_FOR_FLAG(is_subscribed);
+}
+
+static void ccc_changed(const struct bt_gatt_attr *attr, uint16_t value)
+{
+	/* assume we only get it for the `test_gatt_service` */
+	if (value) {
+		SET_FLAG(is_subscribed);
+	}
+}
+
+#define test_service_uuid						\
+	BT_UUID_DECLARE_128(BT_UUID_128_ENCODE(0xf0debc9a, 0x7856, 0x3412, 0x7856, 0x341278563412))
+#define test_characteristic_uuid                                                                   \
+	BT_UUID_DECLARE_128(BT_UUID_128_ENCODE(0xf2debc9a, 0x7856, 0x3412, 0x7856, 0x341278563412))
+
+BT_GATT_SERVICE_DEFINE(test_gatt_service, BT_GATT_PRIMARY_SERVICE(test_service_uuid),
+		       BT_GATT_CHARACTERISTIC(test_characteristic_uuid,
+					      (BT_GATT_CHRC_READ | BT_GATT_CHRC_WRITE |
+					       BT_GATT_CHRC_NOTIFY | BT_GATT_CHRC_INDICATE),
+					      BT_GATT_PERM_READ | BT_GATT_PERM_WRITE,
+					      NULL, NULL, NULL),
+		       BT_GATT_CCC(ccc_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),);
+
+static struct bt_conn *connect_and_subscribe(void)
+{
+	uint16_t handle;
+	struct bt_conn *conn;
+
+	LOG_DBG("Central: Connect to peer");
+	conn = connect_as_central();
+
+	LOG_DBG("Central: Subscribe to peer (conn %p)", conn);
+	find_the_chrc(conn, test_service_uuid, test_characteristic_uuid, &handle);
+	subscribe(conn, handle, notified);
+
+	return conn;
+}
+
+static bool is_disconnected(struct bt_conn *conn)
+{
+	int err;
+	struct bt_conn_info info;
+
+	err = bt_conn_get_info(conn, &info);
+	ASSERT(err == 0, "Failed to get info for %p\n", conn);
+
+	/* Return if fully disconnected */
+	return info.state == BT_CONN_STATE_DISCONNECTED;
+}
+
+static int disconnect(struct bt_conn *conn)
+{
+	if (!conn) {
+		return -ENOTCONN;
+	}
+
+	int err = bt_conn_disconnect(conn, BT_HCI_ERR_REMOTE_USER_TERM_CONN);
+
+	if (!err) {
+		WAIT_FOR_FLAG_UNSET(is_connected);
+	}
+
+	return err;
+}
+
+void entrypoint_dut(void)
+{
+	/* Test purpose:
+	 *
+	 * Verifies that there is no host RX buffer leak due to disconnections.
+	 *
+	 * That is, not actual host buffers (ie memory) but rather the number of
+	 * free buffers that the controller thinks the host has.
+	 *
+	 * If there is a desynchronization between those two, the result is that
+	 * the controller stops forwarding ACL data to the host, leading to an
+	 * eventual application timeout.
+	 *
+	 * To do this, the DUT is connected to a peer that loops through sending
+	 * a few ATT notifications then disconnecting.
+	 *
+	 * The test stops after an arbitrary number of notifications have been
+	 * received.
+	 *
+	 * [verdict]
+	 * - no buffer allocation failures, timeouts or stalls.
+	 */
+	int err;
+	struct dut_state *s = &g_dut_state;
```

Comment:
What is the purpose of accessing the global variable as a pointer, compared to directly accessing it?

Reply:
less typing mostly. It was originally because I previously had two peripherals and had a state object for each of them on the central.